wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/spi/environment/InternalEnvironmentFactory.java

```diff
@@ -78,36 +79,40 @@ public InternalEnvironmentFactory(
    * @throws InfrastructureException if infrastructure specific error occurs
    * @throws ValidationException if validation fails
    */
-  public T create(final Environment sourceEnv) throws InfrastructureException, ValidationException {
+  public T create(@Nullable final Environment sourceEnv)
+      throws InfrastructureException, ValidationException {
 
     Map<String, InternalMachineConfig> machines = new HashMap<>();
     List<Warning> warnings = new ArrayList<>();
+    InternalRecipe recipe = null;
 
-    InternalRecipe recipe = recipeRetriever.getRecipe(sourceEnv.getRecipe());
+    if (sourceEnv != null) {
+      recipe = recipeRetriever.getRecipe(sourceEnv.getRecipe());
 
-    for (Map.Entry<String, ? extends MachineConfig> machineEntry :
-        sourceEnv.getMachines().entrySet()) {
-      MachineConfig machineConfig = machineEntry.getValue();
+      for (Map.Entry<String, ? extends MachineConfig> machineEntry :
+          sourceEnv.getMachines().entrySet()) {
+        MachineConfig machineConfig = machineEntry.getValue();
 
-      List<Installer> installers;
-      try {
-        installers = installerRegistry.getOrderedInstallers(machineConfig.getInstallers());
-      } catch (InstallerException e) {
-        throw new InfrastructureException(e);
+        List<Installer> installers;
+        try {
+          installers = installerRegistry.getOrderedInstallers(machineConfig.getInstallers());
+        } catch (InstallerException e) {
+          throw new InfrastructureException(e);
+        }
+
+        machines.put(
+            machineEntry.getKey(),
+            new InternalMachineConfig(
+                installers,
+                normalizeServers(machineConfig.getServers()),
+                machineConfig.getEnv(),
+                machineConfig.getAttributes(),
+                machineConfig.getVolumes()));
       }
 
-      machines.put(
-          machineEntry.getKey(),
-          new InternalMachineConfig(
-              installers,
-              normalizeServers(machineConfig.getServers()),
-              machineConfig.getEnv(),
-              machineConfig.getAttributes(),
-              machineConfig.getVolumes()));
+      machinesValidator.validate(machines);
     }
 
-    machinesValidator.validate(machines);
-
     return doCreate(recipe, machines, warnings);
```

Comment:
`Recipe` parameter is not marked as `Nullable` but now it is. Please consider adding the corresponding annotations and checks to avoid NPE.

Reply:
Done